deployment/checkreleases.py

```diff
@@ -0,0 +1,64 @@
+from logging import getLogger
+from datetime import timedelta, datetime
+from threading import Condition
+from itertools import repeat
+import json
+
+from .instancehealth import InstanceHealth
+from . import database, samodels as m
+from . import execution, executils
+
+logger = getLogger(__name__)
+
+
+class CheckReleasesWorker(object):
+    """Check releases from every servers and update health info."""
+
+    def __init__(self, frequency, health):
+        logger.info("CheckReleases worker init. It will run every %d seconds" % frequency)
+        self._running = True
+        self._frequency = frequency
+        self._health = health
+        self.wakeup_period = timedelta(seconds=self._frequency)
+        self.condition = Condition()
+
+    def start(self):
+        logger.info("CheckReleases worker wakeup.")
+        self._health.set_ok("releases")
+        while self._running:
+            try:
+                with database.session_scope() as session:
+                    repositories = session.query(m.Repository).all()
+                    for repo in repositories:
+                        for env in repo.environments:
+                            releases = set()
+                            for srv in env.servers:
+                                release = execution.get_release_status(executils.Host.from_server(srv, env.remote_user), env.target_path)
+                                diff_minutes = (datetime.now() - release._release.deployment_date).seconds * 60
```

Comment:
1) Use `datetime.utcnow()` to avoid timezone issues (use the user timezone only when formatting dates, never when storing them - that's the deployer convention anyway, and I found that it greatly simplify datetime management).

2) `_release` is a private attribute (as indicated by the leading underscore), and may be `None` (sometimes release files are not present on servers, for instance during a deployment there is a short time window during which there may be no file. This will abort the whole worker run).
I see several solutions
* add a `get_release` method to `ReleaseStatus` that returns the _release object if success, an raise an exception (gitutils.InvalidReleaseFile for instance) if error
* use gitutils.Release instead (which can only represents success), and add a public function `execution.get_release`  (called by `get_release_status`, and by this worker).

3) `timedelta.seconds` != `timedelta.total_seconds()`  You don't need to convert to seconds anyway, you can compare timedeltas directly (`age = datetime.utcnow() - release._release.deployment_date; if age < timedelta(minutes=30): ...`)

4) You don't handle the case when a server has no release (real life case: I add a server to a cluster, but forget to deploy on it afterwards).